ocp4crd/oauth/oauth.go

```diff
@@ -0,0 +1,135 @@
+package oauth
+
+import (
+	"log"
+
+	configv1 "github.com/openshift/api/legacyconfig/v1"
+	oauthv1 "github.com/openshift/api/oauth/v1"
+	"gopkg.in/yaml.v2"
+
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+func init() {
+	configv1.InstallLegacy(scheme.Scheme)
+	oauthv1.Install(scheme.Scheme)
+}
+
+// TODO: Generated yamls are results of pure imagination. Structure and consistency
+// must be reviewed and fixed. I guess this code can be simplified once we know
+// how the output should look exactly.
+
+// reference:
+//   [v3] OCPv3:
+//   - [1] https://docs.openshift.com/container-platform/3.11/install_config/configuring_authentication.html#identity_providers_master_config
+//   [v4] OCPv4:
+//   - [2] htpasswd: https://docs.openshift.com/container-platform/4.0/authentication/understanding-identity-provider.html
+//   - [3] github: https://docs.openshift.com/container-platform/4.0/authentication/identity_providers/configuring-github-identity-provider.html
+
+// Structures defining custom resource definitions / manifests / yamls
+// TODO: figure out the OKD terminology
+type identityProviderGitHub struct {
+	Name          string `yaml:"name"`
+	Challenge     bool   `yaml:"challenge"`
+	Login         bool   `yaml:"login"`
+	MappingMethod string `yaml:"mappingMethod"`
+	Type          string `yaml:"type"`
+	GitHub        struct {
+		HostName string `yaml:"hostname"`
+		CA       struct {
+			Name string `yaml:"name"`
+		} `yaml:"ca"`
+		ClientID     string `yaml:"clientID"`
+		ClientSecret struct {
+			Name string `yaml:"name"`
+		} `yaml:"clientSecret"`
+		Organizations []string `yaml:"organizations"`
+		Teams         []string `yaml:"teams"`
+	} `yaml:"github"`
+}
+
+type identityProviderHTPasswd struct {
+	Name          string `yaml:"name"`
+	Challenge     bool   `yaml:"challenge"`
+	Login         bool   `yaml:"login"`
+	MappingMethod string `yaml:"mappingMethod"`
+	Type          string `yaml:"type"`
+	HTPasswd      struct {
+		FileData struct {
+			Name string `yaml:"name"`
+		} `yaml:"fileData"`
+	} `yaml:"htpasswd"`
+}
+
+// Shared CRD part, present in all types of OAuth CRDs
+type v4OAuthCRD struct {
+	APIVersion string `yaml:"apiVersion"`
+	Kind       string `yaml:"kind"`
+	MetaData   struct {
+		Name string `yaml:"name"`
+	} `yaml:"metadata"`
+	Spec struct {
+		IdentityProviders []interface{} `yaml:"identityProviders"`
+	} `yaml:"spec"`
+}
+
+// Generate converts OCPv3 OAuth to OCPv4 OAuth Custom Resources
+func Generate(masterconfig configv1.MasterConfig) error {
+	var auth = masterconfig.OAuthConfig
+	var err error
+
+	var crd v4OAuthCRD
+	crd.APIVersion = "config.openshift.io/v1"
+	crd.Kind = "OAuth"
+	crd.MetaData.Name = "cluster"
+
```

Comment:
This stuff above, I guess, is possible to decipher from the original master-config and get rid of hard coding.

Reply:
Yes that's the plan.